### PR TITLE
rgw: ban anonymous user during user creation clearly

### PR DESCRIPTION
**Pull request by mikulely**
uid `anonymous` is reserved by rgw for anonymous user in rgw usage log.

The error msg was confusing, I‘ve encountered this when debug rgw usage log show with `--uid=anonymous`  .

### CODE REVIEW

**Comment by mikulely** [comment]
@joscollin @cbodley mind to review this? Thx!

File: src/rgw/rgw_admin.cc [hunk 1]
```diff
@@ -4333,6 +4333,10 @@ int main(int argc, const char **argv)
   if (!user_id.empty() || !subuser.empty()) {
     ret = user.init(store, user_op);
     if (ret < 0) {
+      if (opt_cmd == OPT_USER_CREATE && user_op.is_preserved_user()) {
+          cerr << "uid: " << user_id << " is preserved by radosgw,"
+	       << " please choose another uid." << std::endl;
+      }
       cerr << "user.init failed: " << cpp_strerror(-ret) << std::endl;
       return -ret;
```
Comment (joscollin) [inline src/rgw/rgw_admin.cc]: @mikulely This is unrelated to your fix: Here `ret` is negative inside `if (ret < 0) {`. But is it making ret positive by putting a `-` sign ? Please ignore this comment if it is intentional.

File: src/rgw/rgw_user.cc [hunk 2]
```diff
@@ -1735,6 +1735,7 @@ int RGWUser::init(RGWUserAdminOpState& op_state)
   }
 
   if (!uid.empty() && (uid.compare(RGW_USER_ANON_ID) != 0)) {
+    op_state.set_preserved_user(false);
```
Comment (mattbenjamin) [inline src/rgw/rgw_user.cc]: does "preserved" mean "reserved" here?

File: src/rgw/rgw_admin.cc (hunk 1 above)
Comment (mikulely) [inline src/rgw/rgw_admin.cc]: I guess here we just following the pattern that returning positive value in the `main`  when found error.

File: src/rgw/rgw_user.cc (hunk 2 above)
Comment (mikulely) [inline src/rgw/rgw_user.cc]: Not a English  native,  it seems "reserved" is more proper.

**Comment by mikulely** [comment]
@joscollin @mattbenjamin followed your suggestion, repushed.

File: src/rgw/rgw_user.cc [hunk 3]
```diff
@@ -1735,6 +1735,7 @@ int RGWUser::init(RGWUserAdminOpState& op_state)
   }
 
   if (!uid.empty() && (uid.compare(RGW_USER_ANON_ID) != 0)) {
+    op_state.set_reserved_user(false);
```
Comment (joscollin) [inline src/rgw/rgw_user.cc]: Here we are checking for the reserved user and avoids the anonymous user. So the value of `reserved_user` should be true logically ?

Comment (mikulely) [inline src/rgw/rgw_user.cc]: As `RGWUserAdminOpState` 's default user is anonymous one(see https://github.com/ceph/ceph/blob/master/src/rgw/rgw_user.h#L478 ), so  `reserved_user` is true by default.

**Comment by mikulely** [comment]
@joscollin please see my comment.(Not sure GitHub would send a notification for single comment in the review, in case Github missed it)

File: src/rgw/rgw_user.cc (hunk 3 above)
Comment (joscollin) [inline src/rgw/rgw_user.cc]: @mikulely Then the `reserved_user` should be `false` right ? I understand that `anonymous user` and `reserved_user` are different.

>     if (!uid.empty() && (uid.compare(RGW_USER_ANON_ID) != 0)) { 
>         op_state.set_reserved_user(false); 

^^ So this is the place the reserved_user becomes true, right ?

Comment (mikulely) [inline src/rgw/rgw_user.cc]: > Then the reserved_user should be false right ?

Yes

>  I understand that anonymous user and reserved_user are different.

Yes, anonymous user is one kind of reserved user. Future more maybe we can make server group user as reserved too.

The `reserved_user` becomes true in `RGWUserAdminOpState` initialization phrase, see
https://github.com/ceph/ceph/pull/16005/files#diff-5ef7b455247dea8b72d475e1ce348284R494

**Comment by mikulely** [comment]
ping @mattbenjamin @joscollin
